src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -238,6 +238,71 @@ impl ScalarFunc {
         let week = t.get_time().iso_week().week();
         Ok(Some(i64::from(week)))
     }
+
+    #[inline]
+    pub fn add_datetime_and_duration<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut t: Cow<'a, Time> = try_opt_or!(
+            self.children[0].eval_time(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let d: Cow<'a, MyDuration> = try_opt_or!(
+            self.children[1].eval_duration(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let add = match t
+            .get_time()
+            .checked_add_signed(Duration::nanoseconds(d.to_nanos()))
+        {
+            Some(result) => result,
+            None => return Err(box_err!("parse from duration {} overflows", d)),
+        };
+        let mut res = t.to_mut().clone();
+        res.set_time(add);
+        Ok(Some(Cow::Owned(res)))
+    }
+
+    #[inline]
+    pub fn add_datetime_and_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut t: Cow<'a, Time> = try_opt_or!(
+            self.children[0].eval_time(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let s: Cow<'a, [u8]> = try_opt_or!(
+            self.children[1].eval_string(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let d = match MyDuration::parse(s.as_ref(), 6) {
```

Comment:
Can you explain the constant "6"?

Reply:
I will replace constant "6" with mysql::MAX_FSP.

Reply:
I see that TiDB use `types.GetFsp()` here, so I think we should use `Time::parse_fsp()` in TiKV instead?

Reply:
Done.